source/common/stats/symbol_table_impl.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/stats/symbol_table.h"
+
+#include "common/common/assert.h"
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Stats {
+
+class SymbolTableImpl : public SymbolTable {
+public:
+  std::vector<Symbol> encode(const std::string& name) override {
+    std::vector<Symbol> symbol_vec;
+    std::vector<std::string> name_vec = absl::StrSplit(name, '.');
+    symbol_vec.resize(name_vec.size());
+    std::transform(name_vec.begin(), name_vec.end(), symbol_vec.begin(),
+                   [this](std::string x) { return this->toSymbol(x); });
+    return symbol_vec;
+  }
+
+  std::string decode(const std::vector<Symbol>& symbol_vec) const override {
+    std::vector<std::string> name;
+    name.resize(symbol_vec.size());
+    std::transform(symbol_vec.begin(), symbol_vec.end(), name.begin(),
```

Comment:
If we care we should micro-benchmark, but keep in mind that StrJoin will right-size the resulting string-array ahead of time, which is better than concatenating iteratively.

Reply:
That makes sense.  However, I think fewer allocations are generally better, and in the current implementation we're already allocating all those bytes in their individual strings within the vector in as well as allocating the vector itself before we even get to allocating the resulting string.  Pre-micro benchmark I think fewer allocations can be assumed to be better until proven otherwise.  

Your point makes sense, though.  If we can put `string_view`s in a preallocated vector and call `StrJoin` on that, it would likely be the best solution since we would get preallocation on both structures with no additional heap use.

Reply:
Q: Before getting into micro perf optimizations, do you have any rough simulations of how much memory this scheme saves? Obviously the symbol table has overhead. I'm just curious where the inflection point is. Is it always better? At what point does it become better? How much better? Etc.